### PR TITLE
Modernizing online_localizer

### PR DESCRIPTION
**Pull request by ovysotska**
This PR modernizes the `online_localizer` and adds the test for it.

### CODE REVIEW

File: src/localization/online_localizer/online_localizer.cpp [hunk 1]
```diff
@@ -32,91 +37,86 @@
...
+OnlineLocalizer::OnlineLocalizer(SuccessorManager *successorManager,
+                                 double expansionRate, double nonMatchingCost) {
+
+  CHECK(successorManager) << "Successor manager is not set.";
+  CHECK(expansionRate > 0 && expansionRate <= 1)
+      << "Graph expansion rate should be in [0,1]. Obtained: " << expansionRate;
+  CHECK(nonMatchingCost > 0)
+      << "Non matching cost should be > 0. Obtained: " << nonMatchingCost;
+
+  successorManager_ = successorManager;
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:82]: It is better to have these as a initializer list, so smth like this:
```c++
OnlineLocalizer::OnlineLocalizer(
  SuccessorManager *successorManager,
  double expansionRate, double nonMatchingCost):
    successorManager_{successorManager},
    expansionRate_{expansionRate},
    nonMatchingCost_{nonMatchingCost} {...}
```

File: src/localization/online_localizer/online_localizer.cpp [hunk 2]
```diff
@@ -32,91 +37,86 @@
...
 
-bool OnlineLocalizer::setNonMatchingCost(double non_match) {
-  // if (non_match < 1.0) {
-  if (non_match < 0.0) {
-    printf("[ERROR][OnlineLocalizer] Invalid Matching cost\n");
-    return false;
-  }
-  _nonMatchCost = non_match;
-  return true;
+OnlineLocalizer::OnlineLocalizer(SuccessorManager *successorManager,
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:73]: Can we make the successor manager const here?

File: src/localization/online_localizer/online_localizer.cpp [hunk 3]
```diff
@@ -32,91 +37,86 @@
...
-    return false;
-  }
-  if (_expansionRate < 0.0) {
-    printf("[ERROR][OnlineLocalizer] Expansion rate is not set\n");
-    return false;
+Matches OnlineLocalizer::findMatchesTill(int queryId) {
+  CHECK(queryId >= 0) << "Number of queries is <= 0: " << queryId;
+  Timer timer;
+  // For the first image consider lost
+  // for every image in the query set
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:98]: Did you by any chance leave the sentence unfinished? Set what?

File: src/localization/online_localizer/online_localizer.cpp [hunk 4]
```diff
@@ -32,91 +37,86 @@
...
+  for (int qu = 0; qu < queryId; ++qu) {
+    // while the graph is not expanded till row 'qu'
+    timer.start();
+    processImage(qu);
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:102]: If this does what the above comment says, maybe a better name for this function would be `processImageUntil` or smth along these lines? Then you wouldn't need a comment.

File: src/localization/online_localizer/online_localizer.cpp [hunk 5]
```diff
@@ -125,110 +125,76 @@ void OnlineLocalizer::writeOutExpanded(const std::string &filename) const {
...
-  _expandedRecently.clear();
+  expandedRecently_.clear();
 
   std::unordered_set<Node> children;
-  if (_needReloc) {
-    _frontier = std::priority_queue<Node>(); // reseting priority_queue
-    printf("[INFO][OnlineLocalizer] RELOCALIZATION\n");
-    Node expandedNode = _currentBestHyp;
+  if (needReloc_) {
+    frontier_ = std::priority_queue<Node>(); // reseting priority_queue
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:141]: It's better to use the uniform initialization here to avoid the ["Most vexing parse"](https://en.wikipedia.org/wiki/Most_vexing_parse):
```c++
frontier_ = std::priority_queue<Node>{};
```

File: src/localization/online_localizer/online_localizer.cpp [hunk 6]
```diff
@@ -125,110 +125,76 @@ void OnlineLocalizer::writeOutExpanded(const std::string &filename) const {
...
+    children.insert(currentBestHyp_);
     updateGraph(expandedNode, children);
   } else {
     bool row_reached = false;
-    printf("[INFO][OnlineLocalizer] NOT LOST\n");
-    while (!_frontier.empty() && !row_reached) {
-      // counterNodes++;
-      Node expandedNode = _frontier.top();
-      _frontier.pop();
+    // printf("[INFO][OnlineLocalizer] NOT LOST\n");
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:158]: Did you forget to remove this comment or do you still need it for something?

File: src/localization/online_localizer/online_localizer.cpp [hunk 7]
```diff
@@ -125,110 +125,76 @@ void OnlineLocalizer::writeOutExpanded(const std::string &filename) const {
...
-    timer.stop();
-    printf("[OnlineLocalizer] Matched image %d\n", qu);
-    timer.print_elapsed_time(TimeExt::MSec);
-    printf("==========================================\n");
-    visualize();
-  }
-  printf("[DEBUG][OnlineLocalizer] Localization finished\n");
-  if (_vis) {
-    _vis->processFinished();
+  // Lost if more than kMaxLostNodesRatio are hidden nodes.
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:192]: nit: I'm not sure we need this comment here.  I think the function name is pretty clear about what it does and how it does it should be clear from the function implementation or documentation.

File: src/localization/online_localizer/online_localizer.cpp [hunk 8]
```diff
@@ -125,110 +125,76 @@ void OnlineLocalizer::writeOutExpanded(const std::string &filename) const {
...
+  if (isLost(kSlidingWindowSize_, kMaxLostNodesRatio)) {
+    needReloc_ = true;
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:194]: nit: I usually prefer longer names. So instead of `needReloc_` I would probably use `needRelocalization_`, but this is of course a matter of personal preference. The same holds for other variables in the code.

File: src/localization/online_localizer/online_localizer.cpp [hunk 9]
```diff
@@ -238,23 +204,19 @@ bool OnlineLocalizer::nodeWorthExpanding(const Node &node) const {
...
 
-  int row_dist = _currentBestHyp.quId - node.quId;
-  if (row_dist < 0) {
-    printf(
-        "[ERROR][INTERNAL][OnlineLocalizer] Trying to expand a node further in "
-        "future %d than current best hypothesis %d.\n",
-        node.quId, _currentBestHyp.quId);
-    exit(EXIT_FAILURE);
-  }
+  int row_dist = currentBestHyp_.quId - node.quId;
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:211]: This can be const. It is not important for performance but it is a nice habit to have, so I will try to point out other places like this too.

File: src/localization/online_localizer/online_localizer.cpp [hunk 10]
```diff
@@ -238,23 +204,19 @@ bool OnlineLocalizer::nodeWorthExpanding(const Node &node) const {
...
+  CHECK(row_dist >= 0) << "Trying to expand a node further in future "
+                       << node.quId << " than current best hypothesis "
+                       << currentBestHyp_.quId;
 
   double mean_cost = computeAveragePathCost();
-  double potential_cost = node.accCost + row_dist * mean_cost * _expansionRate;
+  double potential_cost = node.accCost + row_dist * mean_cost * expansionRate_;
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:217]: This can be `const` too 😉

File: src/localization/online_localizer/online_localizer.cpp [hunk 11]
```diff
@@ -263,21 +225,24 @@ bool OnlineLocalizer::nodeWorthExpanding(const Node &node) const {
...
+    // hypothesis. This is always better.
+    currentBestHyp_ = possibleHyp;
+  } else if (possibleHyp.quId == currentBestHyp_.quId) {
+    // The alternative path to the same level as current best hypothesis as
+    // found.
     double accCost_current =
-        _accCosts[_currentBestHyp.quId][_currentBestHyp.refId];
-    double accCost_poss = _accCosts[possibleHyp.quId][possibleHyp.refId];
+        accCosts_[currentBestHyp_.quId][currentBestHyp_.refId];
+    double accCost_poss = accCosts_[possibleHyp.quId][possibleHyp.refId];
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:238]: This can also be `const` 😉

File: src/localization/online_localizer/online_localizer.cpp [hunk 12]
```diff
@@ -263,21 +225,24 @@ bool OnlineLocalizer::nodeWorthExpanding(const Node &node) const {
...
     if (accCost_poss <= accCost_current) {
-      _currentBestHyp = possibleHyp;
+      // Accumulated cost of new possible hypothesis is smaller than accumulated
+      // cost for the current best hypothesis. This becomes new current best
+      // hypothesis.
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:242]: If you expand the names of your variables the code will say _exactly_ the same thing as your comment does here. I think having longer variables is better for readability and then there will be less need for lengthy comments that tend to go out of sync with the code with time.

File: src/localization/online_localizer/online_localizer.cpp [hunk 13]
```diff
@@ -305,15 +270,15 @@ double OnlineLocalizer::computeAveragePathCost() const {
...
+    pred = pred_.at(pred.quId).at(pred.refId);
   }
   mean_cost = mean_cost / elInPath;
   return mean_cost;
 }
 
 bool OnlineLocalizer::predExists(const Node &node) const {
-  auto quId_found = _pred.find(node.quId);
-  if (quId_found == _pred.end()) {
+  auto quId_found = pred_.find(node.quId);
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:280]: Can also be const :)

File: src/localization/online_localizer/online_localizer.cpp [hunk 14]
```diff
@@ -326,94 +291,77 @@ bool OnlineLocalizer::predExists(const Node &node) const {
 
 void OnlineLocalizer::updateGraph(const Node &parent,
                                   const NodeSet &successors) {
-  if (successors.empty()) {
-    printf(
-        "[WARNING] No successors to add to the graph. May lead to disconnected "
-        "components\n");
-  }
-  // for every successor
+  LOG_IF(WARNING, successors.empty()) << "No successors to add to the graph. "
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:294]: I like the use of `LOG_IF` here 👍

File: src/localization/online_localizer/online_localizer.cpp [hunk 15]
```diff
@@ -326,94 +291,77 @@ bool OnlineLocalizer::predExists(const Node &node) const {
...
-        printf("[DEBUG][OnlineLocalizer] Possible accumulated cost is smaller "
-               "than previous one. Frontier should be updated\n");
-        printf(
-            "[BINGO] If you have reached this line, please contact Olga :)\n");
-        exit(0);
-        // throw 10;
+        LOG(INFO) << "Possible accumulated cost is smaller than previous one. "
+                     "Frontier should be updated";
+        LOG(FATAL)
+            << "[BINGO] If you have reached this line, please contact Olga :)";
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:311]: Maybe it's better to rephrase this to smth like "If this line is printed, please contact Olga" to avoid confusion of people reaching this line when reading the code 😉

File: src/localization/online_localizer/online_localizer.cpp [hunk 16]
```diff
@@ -326,94 +291,77 @@ bool OnlineLocalizer::predExists(const Node &node) const {
...
+        // Update the accumulated cost for a child for estimating the priority
         child.accCost = poss_accCost;
-        // create a child with higher priority ( lower accCost)
-        _frontier.push(child);
+        // Create a child with higher priority ( lower accCost)
+        frontier_.push(child);
       } else {
-        // printf("Child was visited before, but new cost is smaller\n");
-        // child.print();
+        // Child was visited before, but new cost is smaller.
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:322]: Do you need this `else` case?

File: src/localization/online_localizer/online_localizer.cpp [hunk 17]
```diff
@@ -326,94 +291,77 @@ bool OnlineLocalizer::predExists(const Node &node) const {
...
   }
   return path;
 }
 
-/** check starting from the current best node id, if there is a lot of
- * hidden
- * nodes -- meaning LOST **/
-// Check N last image matches. If 90% of them are hidden - LOST!
-
+// Check N last image matches. If `ratio` of them are hidden - LOST!
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:352]: Is this a comment above a doxygen comment? Why not put everything in the doxygen comment?

File: src/localization/online_localizer/online_localizer.cpp [hunk 18]
```diff
@@ -422,9 +370,8 @@ bool OnlineLocalizer::isLost(int N, double perc) const {
       lostFactor++;
     }
   }
-
-  if ((double)lostFactor / path.size() > perc) {
-    printf("[INFO][OnlineLocalizer] LOST localization\n");
+  if ((double)lostFactor / path.size() > ratio) {
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:373]: Don't use the C-style cast as it might use a different cast from what you expect. Use `static_cast<double>(lostFactor)` here instead to guarantee it is using the static cast.

File: src/localization/online_localizer/online_localizer.cpp [hunk 19]
```diff
@@ -456,41 +402,13 @@ std::vector<PathElement> OnlineLocalizer::getLastNmatches(int N) const {
  */
 void OnlineLocalizer::visualize() const {
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.cpp:403]: nit: Probably a better name would be "visualizeIfNeeded" 🤔

File: src/localization/online_localizer/online_localizer.h [hunk 20]
```diff
@@ -85,23 +74,23 @@ class OnlineLocalizer {
   void visualize() const;
 
 private:
-  int _querySize = 0;
-  int _slidingWindowSize = 5; // frames
-  bool _needReloc = false;
-  double _expansionRate = -1.0;
-  double _nonMatchCost = -1.0;
+  const int kSlidingWindowSize_ = 5; // frames
```
Comment (niosus) [inline src/localization/online_localizer/online_localizer.h:77]: I'm not sure if you want to use `const` here. It makes your class non-copyable. If you want this variable to never be changed and to be the same for all instances of this class, it should probably be an: 
```c++
inline static constexpr int kSlidingWindowSize_ = 5;
```
If you want this number to be different for different instances of the class, then you might want to remove `const` altogether and just make sure it does not change using the logic stored in your class.

File: src/localization/successor_manager/node.h [hunk 21]
```diff
@@ -44,11 +44,11 @@ class Node {
   void print() const;
 
   /** WARNING: use only for the priority queue.
- * Comparison is based on the accumulated cost **/
+   * Comparison is based on the accumulated cost **/
```
Comment (niosus) [inline src/localization/successor_manager/node.h:47]: This indentation looks a bit strange. Did you run `clang_format` over this file?

File: src/localization/successor_manager/node.h [hunk 22]
```diff
@@ -44,11 +44,11 @@ class Node {
...
   bool operator<(const Node &rhs) const { return this->accCost > rhs.accCost; }
 };
 
-#define SOURCE_NODE Node(-1, 0, 0.0)
+#define SOURCE_NODE Node(/*queryId=*/-1, /*refId=*/0, /*idvCost=*/0.0)
```
Comment (niosus) [inline src/localization/successor_manager/node.h:51]: Why do you need a macros here? You could instead just create a constant:
```c++
static const Node kConstNode{/*queryId=*/-1, /*refId=*/0, /*idvCost=*/0.0};
```
Or am I missing something here?

File: src/localization/successor_manager/node.h [hunk 23]
```diff
@@ -57,16 +57,15 @@ bool operator==(const Node &lhs, const Node &rhs);
 // custom specialization of std::hash can be injected in namespace std
 // This makes a node hashable to use for std::unordered_set<Node>
 namespace std {
```
Comment (niosus) [inline src/localization/successor_manager/node.h:59]: As of lately the best practices ask not to put anything into the `std` namespace to avoid any form of collision. It would be better to declare a has function here and use it as an explicit parameter to whichever container type that uses it.

File: src/test/online_localizer_test.cpp [hunk 24]
```diff
@@ -0,0 +1,62 @@
+
+#include "database/cost_matrix_database.h"
+#include "features/ifeature.h"
+#include "localization_protos.pb.h"
+#include "online_localizer/online_localizer.h"
+#include "relocalizers/irelocalizer.h"
+#include "successor_manager/successor_manager.h"
+#include "test_utils.h"
```
Comment (niosus) [inline src/test/online_localizer_test.cpp:8]: Do you need all of these includes?

File: src/test/test_utils.h [hunk 25]
```diff
@@ -56,6 +61,27 @@ inline std::filesystem::path createDataForOnlineDatabase() {
...
+
+inline image_sequence_localizer::CostMatrix
+computeCostMatrix(const fs::path &queryDir, const fs::path &refDir) {
+
+  std::vector<std::string> queryFiles = listProtoDir(queryDir, ".Feature");
+  std::vector<std::string> refFiles = listProtoDir(refDir, ".Feature");
+  image_sequence_localizer::CostMatrix cost_matrix;
+
+  for (const auto &query : queryFiles) {
+    auto queryFeature = CnnFeature(query);
```
Comment (niosus) [inline src/test/test_utils.h:73]: Can be const 😉

File: src/test/test_utils.h [hunk 26]
```diff
@@ -56,6 +61,27 @@ inline std::filesystem::path createDataForOnlineDatabase() {
...
+    for (const auto &ref : refFiles) {
+      auto refFeature = CnnFeature(ref);
```
Comment (niosus) [inline src/test/test_utils.h:75]: nit: can be moved to the next line avoiding creating a copy. Not super important here but still a good practice.

File: src/test/test_utils.h [hunk 27]
```diff
@@ -56,6 +61,27 @@ inline std::filesystem::path createDataForOnlineDatabase() {
 inline void clearDataForOnlineDatabase(const std::filesystem::path &path) {
   std::filesystem::remove_all(path);
 }
+
+inline image_sequence_localizer::CostMatrix
+computeCostMatrix(const fs::path &queryDir, const fs::path &refDir) {
+
+  std::vector<std::string> queryFiles = listProtoDir(queryDir, ".Feature");
+  std::vector<std::string> refFiles = listProtoDir(refDir, ".Feature");
```
Comment (niosus) [inline src/test/test_utils.h:69]: Both can be `const`

**Review by niosus — COMMENTED** [review]
Great work! 👏 The code looks much cleaner and even though you've added the whole file of tests + a testing utility you still managed to keep the number of lines added and removed nearly equal. This speaks of the final code being more easy to comprehend, awesome!

Most of my comments are nitpicks, however I did find that sometimes you seem to use comments for fairly straightforward spots of the code which become non-obvious because of the use of shortened names. I believe there is some space for improving readability even further by making the names even more meaningful.

I've only found one place which was a bit strange to me - the use of the macro, but other than that, the PR looks great to me. I'll leave this all as a comment and will approve once the code doesn't have macros in it 😉

File: src/apps/cost_matrix_based_matching/cost_matrix_based_matching_lsh.cpp [hunk 28]
```diff
@@ -85,20 +85,15 @@ int main(int argc, char *argv[]) {
       /*multiProbeLevel=*/2);
   relocalizer->train(loadFeatures(parser.path2ref));
 
-  // initialize SuccessorManager
   std::unique_ptr<SuccessorManager> successorManager =
       std::make_unique<SuccessorManager>(database.get(), relocalizer.get(),
                                          parser.fanOut);
+  online_localizer::OnlineLocalizer localizer(
```
[the remark this answers is not in the document]
Comment (ovysotska) [reply src/apps/cost_matrix_based_matching/cost_matrix_based_matching_lsh.cpp:91]: Actually no, the localizer changes its state when searching for matching images.

File: src/localization/online_localizer/online_localizer.cpp [hunk 29]
```diff
@@ -32,91 +37,86 @@
 #include <limits>
 #include <sstream>
 #include <string>
+#include <unistd.h>
```
[the remark this answers is not in the document]
Comment (ovysotska) [reply src/localization/online_localizer/online_localizer.cpp:40]: Apparently for nothing!

File: src/test/online_localizer_test.cpp (hunk 24 above)
Comment (ovysotska) [reply src/test/online_localizer_test.cpp:8]: It does compile without the includes. But I would like to keep includes since I am using the types from those files.

File: src/test/test_utils.h (hunk 26 above)
Comment (ovysotska) [reply src/test/test_utils.h:75]: Didn't get that.

File: src/test/online_localizer_test.cpp (hunk 24 above)
Comment (niosus) [reply src/test/online_localizer_test.cpp:8]: This is fine. The best practice on this is "include what you see", so if you do use stuff from all of these files they must be included. Thanks!

File: src/test/test_utils.h (hunk 26 above)
Comment (niosus) [reply src/test/test_utils.h]: What I meant is that you can replace this and the following line with:
```c++
cost_matrix.add_values(queryFeature.computeSimilarityScore(CnnFeature(ref)));
```

File: src/localization/successor_manager/node.h (hunk 21 above)
Comment (ovysotska) [reply src/localization/successor_manager/node.h:47]: Yes.
